cl_manycore/libraries/bsg_manycore_cuda.cpp

```diff
@@ -47,312 +56,858 @@ static int hb_mc_get_y(tile_t *tiles, uint8_t *y_list, uint32_t num_tiles) {
-	hb_mc_get_x(tiles, &x_list[0], num_tiles);
-	hb_mc_get_y(tiles, &y_list[0], num_tiles); 
-	hb_mc_load_binary(*fd, elf, &x_list[0], &y_list[0], num_tiles);
-	/* create a memory manager object */
-	if (hb_mc_create_memory_manager(eva_id, elf) != HB_MC_SUCCESS)
-		return HB_MC_FAIL;
-  	
-	/* unfreeze the tile group */
-	for (int i = 0; i < num_tiles; i++) {
-		error = hb_mc_write_tile_reg(*fd, eva_id, &tiles[i], KERNEL_REG, 0x1); /* initialize the kernel register */
-		if (error != HB_MC_SUCCESS)
-			return HB_MC_FAIL;
-		hb_mc_tile_unfreeze_dep(*fd, tiles[i].x, tiles[i].y);
+				bsg_pr_dbg("%s: Grid %d: %dx%d tile group (%d,%d) allocated at origin (%d,%d).\n", __func__, tg-> grid_id, tg->dim.x, tg->dim.y, tg->id.x, tg->id.y, tg->origin.x, tg->origin.y);	
+				return HB_MC_SUCCESS;
+			}
+		}
 	}
-	return HB_MC_SUCCESS;
+	return HB_MC_FAIL;
 }
 
-/*!
- * Initializes Manycore tiles so that they may run kernels.
- * @param fd userspace file descriptor
- * @param eva_id specifies what the EVA-NPA mapping is.
- * @param tiles an array of tile_t structs to initialize.
- * @param num_tiles the number of tiles to initialize.
+
+
+
+/**
+ * Takes the kernel name, argc, argv* and the finish signal address, and initializes a kernel and passes it to tilegroup.
+ * @param[in]  device        Pointer to device
+ * @parma[in]  grid_id       Id of grid to which the tile group belongs
+ * @param[in]  tg_id         Id of tile group
+ * @param[in]  grid_dim      X/Y dimensions of grid to which tile group belongs
+ * @parma[in]  dim           X/Y dimensions of tie pool (mesh)
+ * @param[in]  name          Kernel name that is to be executed on tile group
+ * @param[in]  argc          Number of input arguments to the kernel
+ * @param[in]  argv          List of input arguments to the kernel
  * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
  */
-int hb_mc_device_finish (uint8_t fd, eva_id_t eva_id, tile_t *tiles, uint32_t num_tiles) {
-	if (eva_id != 0) {
-		return HB_MC_FAIL; /* eva_id not supported */
-	} 
+int hb_mc_tile_group_enqueue (device_t* device, grid_id_t grid_id, hb_mc_coordinate_t tg_id, hb_mc_dimension_t grid_dim, hb_mc_dimension_t dim, char* name, uint32_t argc, uint32_t argv[]) {
 
-	else if (!mem_manager[eva_id])
-		return HB_MC_SUCCESS; /* there is no memory manager to deinitialize */
-	
-	delete(mem_manager[eva_id]);
+	if (device->num_tile_groups == device->tile_group_capacity) { 
+		device->tile_group_capacity *= 2;
+		device->tile_groups = (tile_group_t *) realloc (device->tile_groups, device->tile_group_capacity * sizeof(tile_group_t));
+		if (device->tile_groups == NULL) {
+			bsg_pr_err("%s: failed to allocate space for tile_group_t structs.\n", __func__);
+			return HB_MC_FAIL;
+		}
+	}
 	
-	for (int i = 0; i < num_tiles; i++) { /* freeze tiles */
-		hb_mc_tile_freeze_dep(fd, tiles[i].x, tiles[i].y);
+
+	tile_group_t* tg = (tile_group_t *)(device->tile_groups + device->num_tile_groups) ; 
```

Comment:
I recommend avoiding pointer arithmetic. The above line of code is equivalent to this (which is less confusing):
```c 
tile_group_t* tg = &device->tile_groups[device->num_tiles_groups]; 
```

Reply:
Brilliant! fixed, thanks!